File: hosts/android/app/src/main/java/com/dshmobile/spike/CarrierWebDist.kt
package com.dshmobile.spike

import java.io.File
import java.io.OutputStream

/**
 * The fallback seat for the official upstream dist — the frontend-static
 * behavior of docs/webserver-contract.md §2.1 with the Phase-B adaptations
 * of §3.4/§3.6: GET/HEAD only (405 otherwise), traversal → 403, fixed MIME
 * table (+ fonts/png), empty 404s, the auth-lite token→cookie exchange
 * gating index responses only, and the index render pipeline (§1.5):
 * head rows after `<head…>`, body rows after `<body…>`, the
 * `__DSH_BOOT_READY__` tail last, then `<base href="/">`. Kotlin sibling of
 * hosts/ios CarrierWebDist.swift.
 */
class CarrierWebDist(
    private val distRoot: File,
    private val sessionToken: String,
    /** Boot rows, read fresh per render (§1.5: live rows per render). */
    private val indexRows: () -> List<CarrierIndexInjection>,
) {
    /** Evidence hooks — the owning drive logs them; the seat never logs. */
    var onIndexRendered: ((Int, Int) -> Unit)? = null
    var onIndexServed: (() -> Unit)? = null
    var onAssetServed: ((String) -> Unit)? = null

    /** The fallback handler to register on the carrier server. */
    val handler: (CarrierRequest, OutputStream) -> Unit = { request, out -> serve(request, out) }

    private fun serve(request: CarrierRequest, out: OutputStream) {
        if (request.method != "GET" && request.method != "HEAD") {
            return respondEmpty(out, 405) // named routes own their methods
        }
        val path = request.path ?: return respondEmpty(out, 400)
        if (path.contains("..")) return respondEmpty(out, 403)
        if (path == "/" || File(distRoot, path.removePrefix("/")).canonicalPath ==
            File(distRoot, "index.html").canonicalPath
        ) {
            return serveIndex(request, out)
        }
        serveAsset(path, request.method, out)
    }

    // ---- index (auth-lite + render pipeline) --------------------------------

    private fun serveIndex(request: CarrierRequest, out: OutputStream) {
        val token = sessionToken
        if (request.query?.contains("token=$token") == true && request.method == "GET") {
            // mint the session cookie, then bounce to the clean URL (§2.2)
            CarrierHTTP.respond(
                out, 303, "text/plain", ByteArray(0),
                headers = mapOf(
                    "Set-Cookie" to "$COOKIE_HEADER_PREFIX$token; Path=/; HttpOnly",
                    "Location" to "/",
                ),
            )
            return
        }
        if (!request.isAuthed(token)) return respondEmpty(out, 401)
        val index = File(distRoot, "index.html")
        val raw = try {
            index.readText(Charsets.UTF_8)
        } catch (_: Exception) {
            return respondEmpty(out, 404)
        }
        val rows = indexRows()
        val rendered = renderIndex(raw, rows)
        val body = rendered.toByteArray(Charsets.UTF_8)
        onIndexRendered?.invoke(rows.size, body.size)
        onIndexServed?.invoke()
        CarrierHTTP.respond(out, 200, HTML_MIME, body, bodyless = request.method == "HEAD")
    }

    /** Renders rows into the raw index body: head rows after the opening head
     * tag, body rows after the opening body tag, the boot-readiness tail
     * after the last body row (upstream `renderIndexInjections`), then the
     * `<base href="/">` transform immediately after `<head…>` (upstream
     * frontend-static: base lands before the spliced row markup). */
    fun renderIndex(html: String, rows: List<CarrierIndexInjection>): String {
        var head = ""
        var body = ""
        for (row in rows) {
            val (placement, markup) = row.rendered
            if (placement == CarrierIndexInjection.Placement.HEAD) head += markup else body += markup
        }
        body += READY_MARKUP
        var out = html
        out = spliceAfter(out, "<head", "<base href=\"/\">" + head)
        out = spliceAfter(out, "<body", body)
        return out
    }

    companion object {
        val HTML_MIME = "text/html; charset=utf-8"
        const val COOKIE_HEADER_PREFIX = "${CarrierServer.COOKIE_NAME}="

        /** The tail script settling `__DSH_BOOT_READY__` (upstream READY_MARKUP). */
        const val READY_MARKUP =
            "<script>(globalThis.__DSH_BOOT_READY__ ??= Promise.withResolvers()).resolve()</script>"

        /** Inserts `markup` after the opening tag that starts with `pattern`
         * (`<head…>` / `<body…>`); prepend when the document lacks the tag. */
        fun spliceAfter(html: String, pattern: String, markup: String): String {
            val upper = html.lowercase()
            val withBracket = upper.indexOf("$pattern>")
            val at = if (withBracket >= 0) withBracket + pattern.length + 1
            else upper.indexOf(pattern).let { if (it >= 0) it + pattern.length else -1 }
            if (at < 0) return markup + html
            return html.substring(0, at) + markup + html.substring(at)
        }

        /** Upstream's table, extended per §3.6 with fonts and webmanifest icons. */
        val mime = mapOf(
            "html" to HTML_MIME,
            "js" to "text/javascript; charset=utf-8",
            "css" to "text/css; charset=utf-8",
            "svg" to "image/svg+xml",
            "json" to "application/json",
            "map" to "application/json",
            "webmanifest" to "application/manifest+json",
            "gz" to "application/gzip",
            "woff2" to "font/woff2",
            "woff" to "font/woff",
            "ttf" to "font/ttf",
            "png" to "image/png",
        )
    }

    // ---- assets (§2.1 fixed MIME table + §3.6 fonts/png) ----------------------

    private fun serveAsset(path: String, method: String, out: OutputStream) {
        val ext = path.substringAfterLast('.', "").lowercase()
        val contentType = mime[ext] ?: "application/octet-stream"
        val file = File(distRoot, path.removePrefix("/"))
        val data = try {
            if (file.isFile) file.readBytes() else null
        } catch (_: Exception) {
            null
        } ?: return respondEmpty(out, 404) // absent or non-file: empty 404
        onAssetServed?.invoke(path)
        CarrierHTTP.respond(out, 200, contentType, data, bodyless = method == "HEAD")
    }

    private fun respondEmpty(out: OutputStream, status: Int) {
        CarrierHTTP.respond(out, status, "text/plain", ByteArray(0))
    }
}

File: hosts/ios/App/Source/CarrierWebDist.swift
import Foundation
import Network

/// The fallback seat for the official upstream dist — the frontend-static
/// behavior of docs/webserver-contract.md §2.1 with the Phase-B adaptations
/// of §3.4/§3.6: GET/HEAD only (405 otherwise), traversal → 403, fixed MIME
/// table (+ fonts/png), empty 404s, the auth-lite token→cookie exchange
/// gating index responses only, and the index render pipeline (§1.5):
/// head rows after `<head…>`, body rows after `<body…>`, the
/// `__DSH_BOOT_READY__` tail last, then `<base href="/">`.
final class CarrierWebDist {
    /// Evidence hooks — the owning drive logs them; the seat never logs.
    var onIndexRendered: ((Int, Int) -> Void)?
    var onIndexServed: (() -> Void)?
    var onAssetServed: ((String) -> Void)?

    /// Boot rows, read fresh per render (§1.5: live rows per render).
    var indexRows: () -> [CarrierIndexInjection]
    /// Per-session auth-lite token (§3.4: the reduction of §2.2).
    let sessionToken: String
    private let distRoot: URL
    private let distIndex: URL

    init(distRoot: URL, sessionToken: String, indexRows: @escaping () -> [CarrierIndexInjection]) {
        self.distRoot = distRoot
        self.distIndex = distRoot.appendingPathComponent("index.html")
        self.sessionToken = sessionToken
        self.indexRows = indexRows
    }

    /// The fallback handler to register on the carrier server.
    var handler: CarrierHTTPHandler { { [weak self] request, conn in
        self?.serve(request, conn: conn)
    } }

    private func serve(_ request: CarrierRequest, conn: NWConnection) {
        guard request.method == "GET" || request.method == "HEAD" else {
            return respondEmpty(405, conn: conn) // named routes own their methods
        }
        guard let path = request.path else {
            return respondEmpty(400, conn: conn)
        }
        let rel = String(path.dropFirst())
        guard !rel.contains("..") else { return respondEmpty(403, conn: conn) }
        if path == "/" || distRoot.appendingPathComponent(rel).standardizedFileURL
            == distIndex.standardizedFileURL {
            return serveIndex(request, conn: conn)
        }
        serveAsset(path, method: request.method, conn: conn)
    }

    // ---- index (auth-lite + render pipeline) --------------------------------

    private func serveIndex(_ request: CarrierRequest, conn: NWConnection) {
        let token = sessionToken
        if request.query?.contains("token=\(token)") == true, request.method == "GET" {
            // mint the session cookie, then bounce to the clean URL (§2.2)
            let cookie = "\(Self.cookieHeaderPrefix)\(token); Path=/; HttpOnly"
            CarrierServer.respond(status: 303, body: Data(), contentType: "text/plain", conn: conn,
                    headers: ["Set-Cookie": cookie, "Location": "/"])
            return
        }
        guard request.isAuthed(token: token) else {
            return respondEmpty(401, conn: conn)
        }
        guard let raw = try? String(contentsOf: distIndex, encoding: .utf8) else {
            return respondEmpty(404, conn: conn)
        }
        let rows = indexRows()
        let rendered = Self.renderIndex(raw, rows: rows)
        let body = Data(rendered.utf8)
        onIndexRendered?(rows.count, body.count)
        onIndexServed?()
        CarrierServer.respond(status: 200, body: body, contentType: Self.htmlMIME, conn: conn,
                bodyless: request.method == "HEAD")
    }

    /// Renders rows into the raw index body: head rows after the opening head
    /// tag, body rows after the opening body tag, the boot-readiness tail
    /// after the last body row (upstream `renderIndexInjections`), then the
    /// `<base href="/">` transform immediately after `<head…>` (upstream
    /// frontend-static: base lands before the spliced row markup).
    static func renderIndex(_ html: String, rows: [CarrierIndexInjection]) -> String {
        var head = ""
        var body = ""
        for row in rows {
            let rendered = row.rendered
            if rendered.placement == .head { head += rendered.markup }
            else { body += rendered.markup }
        }
        body += readyMarkup
        var out = html
        out = spliceAfter(out, pattern: "<head", markup: "<base href=\"/\">" + head)
        out = spliceAfter(out, pattern: "<body", markup: body)
        return out
    }

    /// The tail script settling `__DSH_BOOT_READY__` (upstream READY_MARKUP).
    static let readyMarkup =
        "<script>(globalThis.__DSH_BOOT_READY__ ??= Promise.withResolvers()).resolve()</script>"

    static let cookieHeaderPrefix = "\(CarrierServer.cookieName)="

    /// Inserts `markup` after the opening tag that starts with `pattern`
    /// (`<head…>` / `<body…>`); prepend when the document lacks the tag.
    static func spliceAfter(_ html: String, pattern: String, markup: String) -> String {
        guard let range = html.range(of: pattern + ">", options: [.caseInsensitive])
            ?? html.range(of: pattern, options: [.caseInsensitive]) else {
            return markup + html
        }
        // Both slices convert explicitly: used directly in `+`, the
        // PartialRangeFrom<String.Index> overload resolves to the
        // Int-bounded subscript on the Xcode 16.4 SDK and fails to compile.
        return String(html[..<range.upperBound]) + markup + String(html[range.upperBound...])
    }

    // ---- assets (§2.1 fixed MIME table + §3.6 fonts/png) ----------------------

    private func serveAsset(_ path: String, method: String, conn: NWConnection) {
        let ext = (path as NSString).pathExtension.lowercased()
        let contentType = Self.mime[ext] ?? "application/octet-stream"
        let file = distRoot.appendingPathComponent(String(path.dropFirst()))
        guard FileManager.default.fileExists(atPath: file.path),
              let data = try? Data(contentsOf: file) else {
            return respondEmpty(404, conn: conn) // absent or non-file: empty 404
        }
        onAssetServed?(path)
        recordAndServe(path, data: data, contentType: contentType,
                       method: method, conn: conn)
    }

    private func recordAndServe(
        _ path: String, data: Data, contentType: String,
        method: String, conn: NWConnection
    ) {
        CarrierServer.respond(status: 200, body: data, contentType: contentType, conn: conn,
                bodyless: method == "HEAD")
    }

    private func respondEmpty(_ status: Int, conn: NWConnection) {
        CarrierServer.respond(status: status, body: Data(), contentType: "text/plain", conn: conn)
    }

    static let htmlMIME = "text/html; charset=utf-8"

    /// Upstream's table, extended per §3.6 with fonts and webmanifest icons.
    static let mime: [String: String] = [
        "html": htmlMIME,
        "js": "text/javascript; charset=utf-8",
        "css": "text/css; charset=utf-8",
        "svg": "image/svg+xml",
        "json": "application/json",
        "map": "application/json",
        "webmanifest": "application/manifest+json",
        "gz": "application/gzip",
        "woff2": "font/woff2",
        "woff": "font/woff",
        "ttf": "font/ttf",
        "png": "image/png",
    ]
}
